src/peer_select.cc

```diff
@@ -582,48 +717,127 @@ PeerSelector::selectSomeNeighbor()
-                                           entry,
-                                           HandlePingReply,
-                                           this,
-                                           &ping.n_replies_expected,
-                                           &ping.timeout);
-
-            if (ping.n_sent == 0)
-                debugs(44, DBG_CRITICAL, "WARNING: neighborsUdpPing returned 0");
-            debugs(44, 3, ping.n_replies_expected <<
-                   " ICP replies expected, RTT " << ping.timeout <<
-                   " msec");
-
-            if (ping.n_replies_expected > 0) {
-                entry->ping_status = PING_WAITING;
-                eventAdd("PeerSelector::HandlePingTimeout",
-                         HandlePingTimeout,
-                         this,
-                         0.001 * ping.timeout,
-                         0);
-                return;
-            }
+    netdbClosestParent(this, request);
+}
+
+static void
+checkNextPingNeighborAccessWrapper(allow_t answer, void *data)
+{
+    PeerSelector *selector = static_cast<PeerSelector *>(data);
+    selector->checkNextPingNeighborAccess(answer);
+}
+
+#if 0
+static void
+selectServersToPingWrapper(PeerSelector *selector)
+{
+    selector->selectServersToPing();
+}
+#endif
+
+void
+PeerSelector::checkNextPingNeighborAccess(allow_t answer)
+{
+    assert(!candidatePingPeers.empty());
+    CbcPointer<CachePeer> p = candidatePingPeers.front();
+    candidatePingPeers.erase(candidatePingPeers.begin());
+
+    if (p.valid())
+        aclPeersCache.push_back(std::pair<CbcPointer<CachePeer>, allow_t>(p, answer));
+
+    if (answer.allowed() && p.valid()) { //check for the next peer
+        debugs(44, 5, "Will ping peer " << p->name);
+        peersToPing.push_back(p);
+    }
+
+#if 0
+    AsyncCall::Pointer call = asyncCall(44, 5, "selectServersToPing", cbdataDialer(selectServersToPingWrapper, this));
+    ScheduleCallHere(call);
+#else
+    selectServersToPing();
+#endif
+}
+
+void
+PeerSelector::selectServersToPing()
+{
+    while (!candidatePingPeers.front().valid())
+        candidatePingPeers.erase(candidatePingPeers.begin());
+
+    if (candidatePingPeers.empty()) { // finish processing candidate peers.
+        doIcpPing();
+        return;
+    }
+
+    CachePeer *p = candidatePingPeers.front().get();
+    if (p->access) {
+        allow_t cached;
+        if (accessCheckCached(p, cached))
+            checkNextPingNeighborAccess(cached);
+        else {
+            ACLFilledChecklist *checklist = new ACLFilledChecklist(p->access, request, NULL);
+            checklist->nonBlockingCheck(checkNextPingNeighborAccessWrapper, this);
         }
+    } else
+        checkNextPingNeighborAccess(ACCESS_ALLOWED);
+}
 
-    if (code != HIER_NONE) {
-        assert(p);
-        addSelection(p, code);
+bool
+PeerSelector::startIcpPing()
+{
+    if (peersToPing.size() > 0 || candidatePingPeers.size() > 0) {
```

Comment:
I am trying to implement your suggested scheme, just I have an objection on this:
```` 
  if (nextPeer.valid() && doIfAllowed(pingAllowedIcpNeighbor, *nextPeer))
            return; // wait for slow cache_peer_access check or a ping response
````
In my patch I am building the list of peers which should be pinked and then ping them together (the doIcpPing). I will keep this because:
 1) The peers responses shows to squid the faster peer and we need to select the faster, so the peers pinging should started more or less at the same time  (we are using first_ping mechanism but it is not enough).
 2) The non-blocking acl checks may take time to complete and this is will influence the results. In my tests just some  debugs are enough to influence the ping results.

I did not finish today, I wil work some hours tomorrow, I hope I will finish it, else I will start working again on Tuesday (It is the Greek Easter here, I will not be able to work the next 3-4 days).

Reply:
If we have to ping all peers at once _and_ we have to check cache_peer_access for peer X before pinging that peer X, then we have to check all cache_peer_access rules (i.e., for all possible peers that are subject to pining). If that logic is correct, then we should check all those rules _before_ we actually start pinging any peers. My continueIcpPing() sketch did not do that, and that was wrong. If your original code did that, please keep that _functionality_, but you still need to rewrite the code so that mortals can understand what is going on. The other parts of my sketch may be useful for that rewrite.

Reply:
I make new changes into this branch.
Please look again in new PeerSelector::selectMore class.
It looks simpler, however still is really complex if you think that it can be called recursively.

With a little work more maybe multiple planNextStep calls can be replaced by one inside PeerSelector::selectMore() function. However we need to add some flags to PeerSelector class to achieve this.  I do not know if it worth to make such  changes.